add methods to export results in tabular format

liannette opened the pull request:
* add method to nplinker class for exporting links, as well as genomic and metabolomic data 
* add test for generating the tabular data for the links

liannette committed:
add print links method to LinkGraph, improve LinkGraph string representation

liannette committed:
feat: add a method to print tabular results files

liannette committed:
improve method names and docstrings, remove unused method to export gcf file

liannette committed:
improve doctring and typing

liannette committed:
fix a failing test

liannette committed:
refactor a little bit the spectrum method to covert to dict

liannette committed:
change the output format for gnps_annotations in metabolomics results file, improve docstrings

liannette committed:
fix: convert int to str before using join

liannette commented:
I thought I was finished, but I realized that it still needs a bit of work. I will assign you as soon as I'm happy with it!

liannette committed:
change representation of empty values in output files for improved integration to excel

liannette committed:
refactoring the export methods

liannette committed:
small refactor: specify staticmethod

liannette committed:
add more tests

liannette committed:
correct typing in doctrings

liannette commented:
@CunliangGeng It's ready for review :) I just can not request a review explicitly, because I have only read permission for the repository.

CunliangGeng reviewed (changes requested):
Let's check the code step by step, and first thing to check is about code format and style:

1. Please check the errors in the [static typing](https://github.com/NPLinker/nplinker/actions/runs/11406802848/job/31871172266?pr=280), and correct the incorrect typings. You could use `mypy` to check them locally before committing.
2. It is not explicitly mentioned, but we do follow some rules for the order of methods/functions in a class/file. The order of methods/functions are:

- `__init__` method
- other magic methods, e.g. `__str__`
- property methods (using `@property`)
- regular methods, class methods (using `@classmethod`), static methods (using `@staticmethod`)
- private methods (`_func` )
- deprecated methods (using `@deprecated`)

For the same level of methods, e.g. regular methods, it's recommended to order them in alphabetical order.

Please check the new methods/functions you added and put them in the right place.

liannette committed:
typing: changed typings to pass mypy static typing checks

liannette committed:
refactor: change the order of methods/functions

CunliangGeng commented on src/nplinker/genomics/bgc.py:
```diff
@@ -173,6 +174,36 @@ def is_mibig(self) -> bool:
         """
         return self.id.startswith("BGC")
 
+    def to_dict(self) -> dict[str, Any]:
```
It's a nice idea to use `to_dict` to output a python class object to a structured and readable format, including but not limited to csv/tsv format. JSON is a another popular format for data exachage. So it's better to make sure the `to_dict` is JSON-compatiable, which means e.g. list should be used over tuple and set (https://docs.python.org/3.10/library/json.html#json.JSONDecoder). Dict could be used as the dict value, but note that the value of the inner dict should not e.g. use set or tuple neither.

Also, make sure that the value of the returned dict does not contain the special characters (e.g. comma, tab) since you want to output them to a csv/tsv.

liannette replied:
Instead of ensuring that the dict values do not contain special characters, I am replacing tabs in the to_tsv methods

CunliangGeng replied:
OK

CunliangGeng commented on src/nplinker/nplinker.py:
```diff
@@ -168,34 +168,50 @@ def scoring_methods(self) -> list[str]:
-        (download, generate and/or validate data) in the [correct directory structure][working-directory-structure],
-        and then calls the [`DatasetLoader`][nplinker.loader.DatasetLoader] class to load all data
-        from the files into memory.
+    def export_objects(self, objects: Sequence[BGC | Spectrum], filename: str) -> None:
```
`objects: Sequence[BGC | Spectrum]` could be a list of mixed BGC and spectrum, so should use `objects: Sequence[BGC]|Sequence[BGC]`.
Also, the code of data validation is needed to make sure all elements of the input list have the same type of data.

CunliangGeng commented on src/nplinker/nplinker.py:
```diff
@@ -168,34 +168,50 @@ def scoring_methods(self) -> list[str]:
-        [`self.bgcs`][nplinker.NPLinker.bgcs] for all BGC objects,
-        [`self.strains`][nplinker.NPLinker.strains] for all Strain objects, etc.
+        Args:
+            objects (list[BGC | Spectrum]): A list of BGC or Spectrum objects to be exported.
```
GCF and MolecularFamily should also be added to this function, and it's necessary to implement their `to_dict` method too.

liannette replied:
I discussed this with Dora, and we both agreed that it is not necessary to export the GCF and MolecularFamily objects, as they do not contain many attributes directly tied to them.

The most important need is to quickly access which BGCs/Spectra are members of which GCF/MolecularFamily, respectively, and this is easiest by filtering the BGC/Spectrum output TSV files. For the GCF, the only interesting attribute is the BiG-SCAPE class, which I instead included as a column when exporting the BGC objects. The MolecularFamily objects do not even contain any interesting attributes.

Do you think it still makes sense to export a file for the GCFs and MolecularFamilies, perhaps for the sake of completeness? I doubt they will get more attributes in the future, but it’s possible.

CunliangGeng replied:
Ok, we don't have to add it if you believe it's not a need for end users.

CunliangGeng commented on src/nplinker/nplinker.py:
```diff
@@ -168,34 +168,50 @@ def scoring_methods(self) -> list[str]:
-        loader.load()
+        headers = objects[0].to_dict().keys()
+        with open(self._output_dir / filename, "w") as f:
+            f.write("\t".join(headers) + "\n")
```
Please use `csv` package to read and write tabluar data.

CunliangGeng commented on src/nplinker/nplinker.py:
```diff
@@ -168,34 +168,50 @@ def scoring_methods(self) -> list[str]:
+                        formatted_row.append("")
+                f.write("\t".join(formatted_row) + "\n")
+
+    def export_results(self, lg: LinkGraph | None = None) -> None:
```
It's better to rename this method to `to_tsv`. And the `export_objects` then renamed to e.g `objects_to_tsv`?

CunliangGeng commented on src/nplinker/scoring/link_graph.py:
```diff
@@ -267,6 +258,101 @@ def filter(self, u_nodes: Sequence[Entity], v_nodes: Sequence[Entity] = [], /) -
+        """
+        return self._g.has_edge(u, v)
+
+    @staticmethod
```
why `staticmethod` is necessary here?

CunliangGeng commented on src/nplinker/scoring/link_graph.py:
```diff
@@ -267,6 +258,101 @@ def filter(self, u_nodes: Sequence[Entity], v_nodes: Sequence[Entity] = [], /) -
+        return self._g.has_edge(u, v)
+
+    @staticmethod
+    def link_to_dict(link: LINK, index: int) -> dict[str, Any]:
```
Keep method names consistent across various classes, so `to_dict` is better.

CunliangGeng replied:
`index` is not necessary for this method, and it can be added to the function which really requires it.

CunliangGeng commented on src/nplinker/scoring/link_graph.py:
```diff
@@ -267,6 +258,101 @@ def filter(self, u_nodes: Sequence[Entity], v_nodes: Sequence[Entity] = [], /) -
+        metabolomic_object = v if isinstance(u, genomic_object_classes) else u
+        metcalf_score = data.get("metcalf")
+        rosetta_score = data.get("rosetta")
+        return {
```
Make sure the returned dict is JSON-compatible.

liannette replied:
The dict values are of types str or float, that's JSON-compatible, right?

CunliangGeng replied:
yep, they are. https://docs.python.org/3.12/library/json.html#json.JSONDecoder

CunliangGeng commented on src/nplinker/scoring/link_graph.py:
```diff
@@ -267,6 +258,101 @@ def filter(self, u_nodes: Sequence[Entity], v_nodes: Sequence[Entity] = [], /) -
+                - rosetta_score (float | str): The Rosetta score, rounded to 2 decimal places.
+        """
+        u, v, data = link
+        genomic_object_classes = (GCF,)
```
Could change the variable name to `genomic_types`

CunliangGeng commented on src/nplinker/scoring/link_graph.py:
```diff
@@ -184,44 +198,21 @@ def add_link(
-        v: Entity,
-    ) -> LINK_DATA | None:
-        """Get the data for a link between two objects.
+    def export_links(self, file: str | PathLike) -> None:
```
Rename it to `to_tsv`?

CunliangGeng commented on src/nplinker/scoring/link_graph.py:
```diff
@@ -184,44 +198,21 @@ def add_link(
-        return self._g.get_edge_data(u, v)  # type: ignore
+        table_data = self.get_table_data()
+        headers = table_data[0].keys()
+        with open(file, "w") as f:
```
Use `csv` to read and write tabular data

CunliangGeng commented on src/nplinker/scoring/link_graph.py:
```diff
@@ -267,6 +258,101 @@ def filter(self, u_nodes: Sequence[Entity], v_nodes: Sequence[Entity] = [], /) -
+        """
+        return self._g.get_edge_data(u, v)  # type: ignore
+
+    def get_table_data(self, display_limit: int | None = None) -> list[dict[str, Any]]:
```
It seems this method can be a private method. 

Also, I prefer renaming it to e.g. `_links_to_dicts`.

CunliangGeng commented on src/nplinker/scoring/link_graph.py:
```diff
@@ -285,35 +371,26 @@ def _filter_two_nodes(self, u: Entity, v: Entity, lg: LinkGraph) -> None:
 
-        if len(self.links) > display_limit:
+        Returns:
+            str: A string representation of the table in GitHub-flavored markdown format. If the number of links
```
The docstring line length should be ≤100, same for other type of lines.

CunliangGeng commented on src/nplinker/scoring/link_graph.py:
```diff
@@ -76,17 +78,17 @@ def __init__(self) -> None:
             >>> lg
-            |    |   Object 1 |   Object 2 |   Metcalf Score |   Rosetta Score |
-            |----|------------|------------|-----------------|-----------------|
+            |    | Genomic Object Type   | Genomic Object ID   | Metabolomic Object Type   | Metabolomic Object ID   | Metcalf Score   | Rosetta Score   |
```
The docstring line length should be ≤100, same for other type of lines. But this docstring can be an exception for clarity.

CunliangGeng reviewed (approved):
Please check the comments in each file.

Last time I mentioned that for the same level of methods, e.g. regular methods, it's recommended to order them in alphabetical order. Well, I forgot to mention that it's better to cluster and order the methods based on their purposes, e.g. it's intuitive to put the input/read/load methods before the output/write/save/dump methods regardless of their names.

Also, please do not change the order of existing methods in this PR, it's better to do it in an independent PR.

CunliangGeng reviewed (changes requested):
This PR needs changes before approval ;-)

liannette commented:
I just looked through your comments, they are quite helpful, thanks!

CunliangGeng commented:
Two more suggestions: 
1. You could use [Github Desktop](https://github.com/apps/desktop) to mange your local commits (e.g. reorder and squash/combine them) with ease. This practise could make the commits more organised, which will make it easier to understand your changes and facilitate the review process.
2. The PR basically could be split to 2 or more smaller PRs to make each PR more atomic and faster to review. For example, the changes to `bgc.py` is independent, which could be done in another PR. You don't have to split this PR now. Next time, when there are independent/irrelevant changes, please consider putting them in another PR.

liannette committed:
restore the order of already existing functions and methods

liannette committed:
make dicts json compatible

liannette committed:
rename functions and variables

liannette committed:
refactor: changed the place when the index is added to the link dict

liannette committed:
use csv package to write the tabular output files

liannette committed:
make sure all elements of the input list have the same type of data.

liannette committed:
shorten to long doc string lines, correct some doc strings

liannette committed:
tests: adapted the test to the changes

liannette committed:
remove a file that was committed by accident

liannette reviewed (commented):
I integrated the requested changes, except the ones in the unresolved conversations. Can you take a look at my replies there?

liannette committed:
Merge branch 'NPLinker:dev' into output_files

liannette commented:
Is there anything more needed before merging?

CunliangGeng commented:
> Is there anything more needed before merging?

I'll take a closer look today.

CunliangGeng commented on src/nplinker/genomics/bgc.py:
```diff
@@ -173,6 +174,36 @@ def is_mibig(self) -> bool:
         return self.id.startswith("BGC")
 
+    def to_dict(self) -> dict[str, Any]:
+        """Convert the BGC object to a dictionary for exporting results.
```
```suggestion
        """Convert the BGC object to a dictionary for exporting purpose.
```

CunliangGeng commented on src/nplinker/genomics/bgc.py:
```diff
@@ -173,6 +174,36 @@ def is_mibig(self) -> bool:
+
+        Compiles relevant information from the BGC object and formats it into a dictionary.
+        Each key-value pair in the dictionary represents a specific attribute of the BGC.
+
```
```suggestion
```

CunliangGeng commented on src/nplinker/genomics/bgc.py:
```diff
@@ -173,6 +174,36 @@ def is_mibig(self) -> bool:
+        Each key-value pair in the dictionary represents a specific attribute of the BGC.
+
+        Returns:
+            A dictionary containing the following key-value pairs:
```
```suggestion
            A dictionary containing the following key-value pairs:

```

Add one blank line to ensure the following list can be properly rendered for doc website.

CunliangGeng commented on src/nplinker/genomics/bgc.py:
```diff
@@ -173,6 +174,36 @@ def is_mibig(self) -> bool:
+        Returns:
+            A dictionary containing the following key-value pairs:
+            - GCF_id (list[str]): A list of GCF IDs.
+            - GCF_bigscape_class (list[str | None]): A list of BiG-SCAPE classes.
```
```suggestion
            - GCF_bigscape_class (list[str]): A list of BiG-SCAPE classes.
```
Please check if it's what you want.

CunliangGeng commented on src/nplinker/genomics/bgc.py:
```diff
@@ -173,6 +174,36 @@ def is_mibig(self) -> bool:
+            "mibig_bgc_class": self.mibig_bgc_class,
+            "antismash_id": self.antismash_id,
+            "antismash_region": self.antismash_region,
+        }
```
Please reorder the dict keys to make the structure logical, e.g. you could follow how this BGC class orgarnises these info.

After that, you also need to update the docstring.

liannette replied:
I arranged the dictionary keys to create an optimal column order for the end user when reviewing the results, so I’d like to preserve this arrangement.

CunliangGeng replied:
OK, please add a comment to mention it.

CunliangGeng commented on src/nplinker/metabolomics/spectrum.py:
```diff
@@ -97,3 +98,29 @@ def has_strain(self, strain: Strain) -> bool:
         return strain in self.strains
+
+    def to_dict(self) -> dict[str, Any]:
+        """Convert the Spectrum object to a dictionary for exporting results.
```
```suggestion
        """Convert the Spectrum object to a dictionary for exporting purpose.
```

CunliangGeng commented on src/nplinker/metabolomics/spectrum.py:
```diff
@@ -97,3 +98,29 @@ def has_strain(self, strain: Strain) -> bool:
+
+        This method compiles relevant information from the Spectrum object into a dictionary format.
+        Each key-value pair in the dictionary represents a specific attribute of the Spectrum Object.
+
```
```suggestion
```

CunliangGeng commented on src/nplinker/metabolomics/spectrum.py:
```diff
@@ -97,3 +98,29 @@ def has_strain(self, strain: Strain) -> bool:
+        Each key-value pair in the dictionary represents a specific attribute of the Spectrum Object.
+
+        Returns:
+            A dictionary containing containing the following key-value pairs:
```
```suggestion
            A dictionary containing containing the following key-value pairs:

```

CunliangGeng commented on src/nplinker/nplinker.py:
```diff
@@ -355,3 +356,55 @@ def save_data(
             pickle.dump(data, f)
+
+    def objects_to_tsv(self, objects: Sequence[BGC] | Sequence[Spectrum], filename: str) -> None:
+        """Exports a list of BGC or Spectrum objects to a specified file in tab-separated format.
```
```suggestion
        """Exports a list of BGC or Spectrum objects to a tsv file.
```

CunliangGeng commented on src/nplinker/nplinker.py:
```diff
@@ -355,3 +356,55 @@ def save_data(
+
+        Args:
+            objects (list): A list of BGC or a list of Spectrum objects to be exported.
+            filename (str): The name of the file where the data will be saved.
```
```suggestion
            filename (str): The name of the output file.
```

CunliangGeng commented on src/nplinker/nplinker.py:
```diff
@@ -355,3 +356,55 @@ def save_data(
+            raise TypeError("All objects in the list must be of the same type")
+
+        headers = objects[0].to_dict().keys()
+        with open(self._output_dir / filename, "w", newline="") as outfile:
```
```suggestion
        with open(self._output_dir / filename, "w") as outfile:
```

liannette replied:
See the other comment about the necessity of the newline="". It is necessary to prevent extra blank lines.

CunliangGeng commented on src/nplinker/nplinker.py:
```diff
@@ -355,3 +356,55 @@ def save_data(
+                    else:
+                        row[header] = str(value) if value else ""
+                    # Replace tabs with 4 spaces
+                    row[header] = row[header].replace("\t", "    ")
```
This part could be extracted to a `to_tabular` method with some modifications for both BGC and Spectrum classes.

Then you could use pandas dataframe to faciliate the process of a list of objects.

liannette replied:
I moved this part into a `to_string` method. Let me know if that's okay as well.

CunliangGeng commented on src/nplinker/nplinker.py:
```diff
@@ -355,3 +356,55 @@ def save_data(
+                writer.writerow(row)
+
+    def to_tsv(self, lg: LinkGraph | None = None) -> None:
+        """Exports the results to the output directory in tab-separated format.
```
```suggestion
        """Export data to tsv files.
```

CunliangGeng commented on src/nplinker/nplinker.py:
```diff
@@ -355,3 +356,55 @@ def save_data(
+    def to_tsv(self, lg: LinkGraph | None = None) -> None:
+        """Exports the results to the output directory in tab-separated format.
+
+        This method exports genomics and metabolomics data to their respective
+        TSV files in the specified output directory. If a LinkGraph object is
+        provided, it also exports the links data to a TSV file.
```
```suggestion
        This method exports following data to seperated TSV files:
        
         - BGC objects: `genomics_data.tsv`
         - Spectrum objects: `metabolomics_data.tsv`
         - LinkGraph object (if given): `links.tsv`
```

CunliangGeng commented on src/nplinker/scoring/link_graph.py:
```diff
@@ -267,6 +282,53 @@ def filter(self, u_nodes: Sequence[Entity], v_nodes: Sequence[Entity] = [], /) -
+            link: A tuple containing the link information (u, v, data).
+
+        Returns:
+            A dictionary containing the link information with the following keys:
```
```suggestion
            A dictionary containing the link information with the following keys:
            
```

CunliangGeng commented on src/nplinker/scoring/link_graph.py:
```diff
@@ -267,6 +282,53 @@ def filter(self, u_nodes: Sequence[Entity], v_nodes: Sequence[Entity] = [], /) -
+        }
+
+    def to_tsv(self, file: str | PathLike) -> None:
+        """Exports the links in the LinkGraph to a file in tab-separated format.
```
```suggestion
        """Exports the links in the LinkGraph to a TSV file.
```

CunliangGeng commented on src/nplinker/scoring/link_graph.py:
```diff
@@ -267,6 +282,53 @@ def filter(self, u_nodes: Sequence[Entity], v_nodes: Sequence[Entity] = [], /) -
+        """Exports the links in the LinkGraph to a file in tab-separated format.
+
+        Args:
+            file: the file to write the links to.
```
```suggestion
            file: the path to the output TSV file.
```

CunliangGeng commented on src/nplinker/scoring/link_graph.py:
```diff
@@ -267,6 +282,53 @@ def filter(self, u_nodes: Sequence[Entity], v_nodes: Sequence[Entity] = [], /) -
+        """
+        table_data = self._links_to_dicts()
+        headers = table_data[0].keys()
+        with open(file, "w", newline="") as f:
```
why is `newline=""` necessary?

liannette replied:
The csv module manages its own newlines internally. If newline="" is not specified, it results in blank lines in the output file, as the csv.DictWriter adds its own line endings on top of those added by open.

CunliangGeng commented on src/nplinker/scoring/link_graph.py:
```diff
@@ -285,35 +347,47 @@ def _filter_two_nodes(self, u: Entity, v: Entity, lg: LinkGraph) -> None:
-            metcalf_score = data.get("metcalf")
-            rosetta_score = data.get("rosetta")
+        Returns:
+            str: A string representation of the table in GitHub-flavored markdown format. If the
```
```suggestion
           A string representation of the table in GitHub-flavored markdown format. If the
```

CunliangGeng reviewed (changes requested):
Almost there! A few more minor comments to resolve.

liannette committed:
Improve docstrings

Apply suggestions from code review

Co-authored-by: Cunliang Geng <c.geng@esciencecenter.nl>

liannette committed:
Improve docstrings

Apply suggestions from code review

Co-authored-by: Cunliang Geng <c.geng@esciencecenter.nl>

liannette committed:
refactor: add method to convert a value to string for tabular output

liannette committed:
Merge branch 'output_files' of https://github.com/liannette/nplinker into output_files

CunliangGeng commented on src/nplinker/nplinker.py:
```diff
@@ -355,3 +356,68 @@ def save_data(
+                row = obj.to_dict()
+                for key, value in row.items():
+                    row[key] = self.to_string(value).replace("\t", "    ")
+                writer.writerow(row)
```
```suggestion
            for obj in objects:
                writer.writerow(obj.to_tabular())
```

It's better to implement the transformation in each object class, i.e. `to_tabluar` (or your `to_string`), which is more modular. So that wherever it's needed, you just call it instead of implementing it again.

CunliangGeng reviewed (changes requested):
One more comment ;-)

liannette committed:
improve doctring, add a comment about key order of bgc dict representation

liannette committed:
move to_string method to the BGC/Spectrum class, add a to_tabular method

liannette committed:
add tests for the to_string method

liannette committed:
change to_tabular to it returns a list and not a string

CunliangGeng commented on src/nplinker/genomics/bgc.py:
```diff
@@ -201,6 +203,36 @@ def to_dict(self) -> dict[str, Any]:
         }
 
+    @staticmethod
+    def to_string(value: Any) -> str:
```
```suggestion
    def _to_string(value: Any) -> str:
```

CunliangGeng commented on src/nplinker/genomics/bgc.py:
```diff
@@ -201,6 +203,36 @@ def to_dict(self) -> dict[str, Any]:
+            A string representation of the input value.
+        """
+        # Convert list, tuple, set to comma-separated string
+        if isinstance(value, (list, tuple, set)):
```
Actually you will never have tuple or set if your dict is JSON-compatible

CunliangGeng commented on src/nplinker/genomics/bgc.py:
```diff
@@ -201,6 +203,36 @@ def to_dict(self) -> dict[str, Any]:
+        # Convert dict to comma-separated string
+        elif isinstance(value, dict):
+            value = ", ".join([f"{k}:{v}" for k, v in value.items()])
+        # Convert anything else to string
```
Do you really want to covert `None` to ""? Does it make sense to the BGC attributes?

liannette replied:
If the value is `None`, the corresponding field in the tabular output file should be left empty. This ensures that when the file is opened in Excel, numeric fields are correctly recognized as numbers rather than text, allowing the columns to be sorted properly. For text fields, leaving them empty is also preferable to displaying None, as it is cleaner and more intuitive.

CunliangGeng commented on src/nplinker/genomics/bgc.py:
```diff
@@ -201,6 +203,36 @@ def to_dict(self) -> dict[str, Any]:
+            value = str(value) if value else ""
+        return value
+
+    def to_tabular(self) -> list:
```
It's better to return a `dict` instead of a `list`, so that you have both the keys and values, making the use easier.
Otherwise, you have to call `to_dict()` to get the keys, and worry about the order of keys or values.

CunliangGeng commented on src/nplinker/metabolomics/spectrum.py:
```diff
@@ -122,3 +122,33 @@ def to_dict(self) -> dict[str, Any]:
         }
+
+    @staticmethod
+    def to_string(value: Any) -> str:
```
Please see my comments in the `BGC.py` and do same changes for this file

CunliangGeng commented on src/nplinker/nplinker.py:
```diff
@@ -374,41 +374,17 @@ def objects_to_tsv(self, objects: Sequence[BGC] | Sequence[Spectrum], filename:
-                row = obj.to_dict()
-                for key, value in row.items():
-                    row[key] = self.to_string(value).replace("\t", "    ")
+                row = [item.replace("\t", "    ") for item in obj.to_tabular()]
```
The replace of `\t` can be done in the `to_tablular` method.

CunliangGeng reviewed (changes requested):
A few more comments, almost there!

liannette committed:
refactor: to_tabular returns dict, to_string turned into private func, tabs are replaced in to_tabular

liannette commented:
Alright, I really hope that we can merge this now 👀

CunliangGeng commented on src/nplinker/genomics/bgc.py:
```diff
@@ -225,13 +225,16 @@ def to_string(value: Any) -> str:
         return value
 
-    def to_tabular(self) -> list:
+    def to_tabular(self) -> dict[str, Any]:
```
```suggestion
    def to_tabular(self) -> dict[str, str]:
```

CunliangGeng commented on src/nplinker/genomics/bgc.py:
```diff
@@ -204,18 +204,18 @@ def to_dict(self) -> dict[str, Any]:
         Args:
             value: The value to be converted to a string.
-                Can be a list, tuple, set, dict, or any other type.
+                Can be a list, dict, or any other type.
```
```suggestion
                Can be a list, dict, or other JSON-compatible types.
```

CunliangGeng commented on src/nplinker/genomics/bgc.py:
```diff
@@ -204,18 +204,18 @@ def to_dict(self) -> dict[str, Any]:
         Returns:
             A string representation of the input value.
         """
         # Convert list, tuple, set to comma-separated string
```
```suggestion
        # Convert list to comma-separated string
```

CunliangGeng commented on src/nplinker/genomics/bgc.py:
```diff
@@ -225,13 +225,16 @@ def to_string(value: Any) -> str:
 
         Returns:
-            list: A list of strings representing the BGC object in tabular format.
+            dict: A dictionary representing the BGC object in tabular format.
```
```suggestion
            dict: A dictionary representing the BGC object in tabular format.
                The keys can be treated as headers and values are strings in which tabs are removed.
                This dict can be exported as a TSV file.
```

CunliangGeng commented on tests/unit/genomics/test_bgc.py:
It's required to test public methods `to_tabluar` and optional to test private method `_to_string`. So please add unit tests for the public methods.

CunliangGeng commented on tests/unit/metabolomics/test_spectrum.py:
It's required to test public methods `to_tabluar` and optional to test private method `_to_string`. So please add unit tests for `to_tabluar` method.

CunliangGeng commented on src/nplinker/metabolomics/spectrum.py:
Please update the docstrings and comments based on my comments in `bgc.py`

CunliangGeng commented on tests/unit/scoring/test_link_graph.py:
Please add unit tests for public method `to_tsv`.

CunliangGeng commented on src/nplinker/nplinker.py:
Please add unit tests for the new public methods `objects_to_tsv` and `to_tsv` in `test_nplinker_local.py`.

CunliangGeng reviewed (changes requested):
I also hope we can merge it now :shipit:  However, we must adhere to some important quality requirements, such as making docstrings and comments accurate and clear; and adding unit tests for all public unit tests. These quality requirements will benefit everyone over time, rather than accumulating more and more technical debt

You could choose what to do next: 
1 🍎 Continue this PR and resolve the remaining comments 
2 🍐 Merge this PR and open some new and small PRs to resolve the remaining comments

Let me know which one you prefer.

liannette committed:
fix typing in to_tabular methods

liannette committed:
update docstrings and comments

liannette committed:
ensure 0 and 0.0 are correctly converted to strings, and not to empty strings

liannette committed:
change the order of methods

liannette committed:
remove whitespace in blank lines

liannette committed:
update and add tests

liannette committed:
change variable name to fix mypy error

liannette committed:
test: trying to fix unit test issue where the spectrum rt is a dict instead of numerical

liannette committed:
Merge branch 'NPLinker:dev' into output_files

liannette committed:
tests: add precursor charge to the test spectra

liannette commented:
Fair enough! I added the unit tests for the public methods/functions and updated the comments and doc strings. Thanks for the detailed suggestions, I learned a lot, even if it's a lot of effort.

If you have any more comments, I would prefer to 🍐 merge this PR and open smaller PRs to resolve them, if possible. You'll need to merge, as I have write access.

CunliangGeng commented on src/nplinker/metabolomics/spectrum.py:
```diff
@@ -134,35 +134,40 @@ def to_dict(self) -> dict[str, Any]:
+        """Convert the Spectrum object to a tabular format.
+
+        Returns:
+            dict: A dictionary representing the BGC object in tabular format.
```
```suggestion
            dict: A dictionary representing the Spectrum object in tabular format.
```

CunliangGeng commented on tests/unit/metabolomics/test_spectrum.py:
```diff
@@ -103,9 +103,45 @@ def __init__(self, id):
+    assert tabular_repr["gnps_annotations"] == ""
+
+    # Test with molecular family
+    class MockMolecularFamily:
```
Why not using `MolecularFamily` directly?

CunliangGeng reviewed (approved):
🎉 It's ready to merge!!!

One minor comment is that why you use `MockMolecularFamily` in the tests instead of the real `MolecularFamily`. You could solve it in a new PR.

CunliangGeng committed:
Update src/nplinker/metabolomics/spectrum.py
